apps/hash-frontend/src/components/hooks/use-agent-runner.ts

```diff
@@ -24,20 +27,25 @@ export const useAgentRunner = <T extends AgentType["Agent"]>(
   });
 
   const callAgentRunnerCallback = useCallback(
-    async (input: Extract<AgentType, { Agent: T }>["Input"]) => {
-      return (
-        await callAgentRunnerFn({
-          variables: {
-            payload: {
-              Agent: agent,
-              Input: input,
-            },
-          },
-        })
-      ).data?.callAgentRunner.Output;
+    async (input: Parameters<CallAgentRunnerCallback<T>>[0]) => {
+      const { data, errors } = await callAgentRunnerFn({
+        variables: {
+          payload: { Agent: agent, Input: input },
+        },
+      });
+
+      const output = data?.callAgentRunner.Output as Extract<
+        AgentType,
+        { Agent: T }
+      >["Output"];
+
+      return { output, errors };
     },
     [agent, callAgentRunnerFn],
   );
 
-  return [callAgentRunnerCallback, { loading }];
+  return [
+    callAgentRunnerCallback as unknown as CallAgentRunnerCallback<T>,
```

Comment:
@thehabbos007 might be able to provide more detail, as he worked on it last week, but without it there's the following type mismatch between the `input` of the method and the `input` of the generated types:
```
Type '(input: Parameters<CallAgentRunnerCallback<T>>[0]) => Promise<{ output: Output | Output | undefined; errors: readonly GraphQLError[] | undefined; }>' is not assignable to type 'CallAgentRunnerCallback<T>'.
  Types of parameters 'input' and 'input' are incompatible.
    Type 'Extract<{ Agent: "generate-entity-type-property-types"; Input: Input; Output: Output; }, { Agent: T; }>["Input"] | Extract<{ Agent: "template"; Input: Input; Output: Output; }, { ...; }>["Input"]' is not assignable to type 'Parameters<CallAgentRunnerCallback<T>>[0]'.
      Type 'Extract<{ Agent: "generate-entity-type-property-types"; Input: Input; Output: Output; }, { Agent: T; }>["Input"]' is not assignable to type 'Parameters<CallAgentRunnerCallback<T>>[0]'.
        Type 'Extract<{ Agent: "generate-entity-type-property-types"; Input: Input; Output: Output; }, { Agent: T; }>' is not assignable to type 'Parameters<CallAgentRunnerCallback<T>>'.
          Type '{ Agent: T; } & { Agent: "generate-entity-type-property-types"; Input: Input; Output: Output; }' is not assignable to type 'Parameters<CallAgentRunnerCallback<T>>'.ts(2322)
io_types.ts(11, 3): 'expression' is declared here.
```

I spent a while trying to figure out a cleaner solution, but couldn't come up with one that didn't involve hardcoding the agent types in this file (which is something we wanted to avoid).

Reply:
I was able to make some progress on isolating the specific type issue and confirm it's TS being bad, and not an actual type error. I still couldn't quite figure out what's going on, but this at least allows us to use `@ts-expect-error` on the actual error, instead of just casting and potentially covering up other issues.

See: https://github.com/hashintel/hash/compare/bw/ai-property-types...nh/type-cast-investigation

Reply:
Thanks Nate for your help investigating, prefer what you've done so have cherry-picked your commit in https://github.com/hashintel/hash/pull/2486/commits/3d56a1085d18ab864471580738f7dd1f8a68b5d6.